sdk-java/src/main/java/ly/count/sdk/java/internal/ModuleBackendMode.java

```diff
@@ -0,0 +1,434 @@
+                return;
+            }
+
+            StringWriter sw = new StringWriter();
+            PrintWriter pw = new PrintWriter(sw);
+            throwable.printStackTrace(pw);
+
+            recordException(deviceID, throwable.getMessage(), sw.toString(), segmentation, timestamp);
+        }
+
+        public void recordException(String deviceID, String message, String stacktrace, Map<String, String> segmentation, long timestamp) {
+            if (!internalConfig.isBackendModeEnable()) {
+                L.wtf("[Countly][BackendMode][recordException] BackendMode is not enable.");
+                return;
+            }
+
+            if (deviceID == null || deviceID.isEmpty()) {
+                L.wtf("[Countly][BackendMode][recordException] DeviceID can not be null or empty.");
+                return;
+            }
+            if (message == null || message.isEmpty()) {
+                L.wtf("[Countly][BackendMode][recordException] message can not be null or empty.");
+                return;
+            }
+
+            if (stacktrace == null) {
+                L.wtf("[Countly][BackendMode][recordException] stacktrace can not be null.");
+                return;
+            }
+
+            JSONObject crash = new JSONObject();
+            crash.put("_error", stacktrace);
+            crash.put("_custom", segmentation);
+            crash.put("_name", message);
+            //crash.put("_nonfatal", true);
+
+            Request request = new Request();
+            request.params.add("device_id", deviceID);
+            request.params.add("crash", crash);
+
+            addTimeInfoIntoRequest(request, timestamp < 1 ? System.currentTimeMillis() : timestamp);
+
+            requestQ.add(request);
+        }
+
+        public void recordUserProperties(String deviceID, Map<String, Object> userProperties, long timestamp) {
+            if (!internalConfig.isBackendModeEnable()) {
+                L.wtf("[Countly][BackendMode][recordUserProperties] BackendMode is not enable.");
+                return;
+            }
+
+            if (deviceID == null || deviceID.isEmpty()) {
+                L.wtf("[Countly][BackendMode][recordUserProperties] DeviceID can not be null or empty.");
+                return;
+            }
+            if (userProperties == null || userProperties.isEmpty()) {
+                L.wtf("[Countly][BackendMode][recordUserProperties] userProperties can not be null or empty.");
+                return;
+            }
+
+            Request request = new Request();
+            JSONObject properties = new JSONObject(userProperties);
+
+            request.params.add("device_id", deviceID);
+            request.params.add("user_details", properties);
+
+            addTimeInfoIntoRequest(request, timestamp < 1 ? System.currentTimeMillis() : timestamp);
+            requestQ.add(request);
+        }
+
+        public ModuleBase getModule() {
```

Comment:
Why does this exist?

Reply:
We are using to get instance of outclass ''ModuleBackendMode" and use it fields (eventQ, requestQ etc...)

Reply:
Bad design. It should not be part of the public API